db.py

```diff
@@ -48,6 +51,16 @@ def allowed(self, access_level):
         else:
             return False
 
+    @staticmethod
+    def validate_unique_email(mapper, connect, target):
+        if target.email:
+            query = db.session.query(User).filter(User.email == target.email, User.id != target.id).exists()
+            if db.session.query(query).scalar():
+                raise ValueError("Email already exists")
+
+
+event.listen(User, 'before_insert', User.validate_unique_email)
```

Comment:
Defining a hook is useful in case you don't exactly know where the insertions will be done, or in case there are several known locations.

In the context of this app user insertion is done at two locations : when the user has logged in by themselve, or when they are added by the administrators. In the first case, the check is performed to know if the account must be created. So there is only one location you may have a conflict.

Therefore, maybe it is more straightforward and **explicit** to perform this check in the appropriate route.

According to the User model definition, the only integrity error that may arise is a duplicate email. You can also fully rely on exception handling and return a comprehensive message to the user in case an IntegrityError has been raised. This is not defensive but not wrong neither, as it is just the propagated result of the DBMS inner check.